src/ucs/stats/stats.c

```diff
@@ -477,95 +481,108 @@ void ucs_stats_node_free(ucs_stats_node_t *node)
-                        .class_name,
-                sizeof(ucs_stats_context
-                               .aggrgt_sum_db[ucs_stats_context.aggrgt_sum_db_size]
-                               .class_name),
-                "%s", class_name);
-        ucs_stats_context.aggrgt_sum_db[ucs_stats_context.aggrgt_sum_db_size]
-                .aggregate_sum_class_offset =
-                ucs_stats_context.aggrgt_sum_db_last_offset;
-
-        /* Initialize return value */
-        offset = ucs_stats_context.aggrgt_sum_db_last_offset;
-
-        /* Return class_id */
-        *class_id = ucs_stats_context.aggrgt_sum_db_size;
-
-        /* Update global offset */
-        ucs_stats_context.aggrgt_sum_db_last_offset += num_counters_in_class;
-
-        /* Reallocate counters names db */
-        ucs_stats_context.aggrgt_cnt_names_db_size += num_counters_in_class;
-        total_alloc_size_bytes =
-                (ucs_stats_context.aggrgt_cnt_names_db_size *
-                 sizeof(*ucs_stats_context.aggrgt_cnt_names_db));
-        ucs_stats_context.aggrgt_cnt_names_db = ucs_realloc(
-                ucs_stats_context.aggrgt_cnt_names_db, total_alloc_size_bytes,
-                (const char*)UCS_STATS_AGGRGT_SUM_CNT_NAMES_DB);
-
-        cnt_agrgt_sum_index = offset;
-        ucs_for_each_bit(counter_index, filter_node->counters_bitmask) {
-            ucs_list_for_each(temp_node, &filter_node->type_list_head,
-                              type_list) {
-                ucs_stats_context.aggrgt_cnt_names_db[cnt_agrgt_sum_index].name =
-                        node->cls->counter_names[counter_index];
-                ucs_stats_context.aggrgt_cnt_names_db[cnt_agrgt_sum_index]
-                        .class_name = node->cls->name;
-            }
-            cnt_agrgt_sum_index++;
+        capacity = ucs_array_capacity(&ucs_stats_context.aggrgt_sum_db);
+        ucs_array_set_length(&ucs_stats_context.aggrgt_sum_db, capacity);
+    }
+
+    element = &ucs_array_elem(&ucs_stats_context.aggrgt_sum_db, db_size);
+    printf("element = %p\n", element);
+
+    /* Initialize entry */
+    ucs_snprintf_safe(element->class_name, sizeof(element->class_name),
+                      "%s", class_name);
+    element->aggregate_sum_class_offset = ucs_stats_context.aggrgt_sum_db_last_offset;
+
+    /* Initialize return value */
+    offset = ucs_stats_context.aggrgt_sum_db_last_offset;
+
+    /* Return class_id */
+    node->cls->class_id = db_size;
+
+    /* Update global offset */
+    ucs_stats_context.aggrgt_sum_db_last_offset += n_cnts_in_cls;
+
+    /* Reallocate counters names db */
+    db_size = ucs_stats_context.aggrgt_cnt_names_db_size += n_cnts_in_cls;
+
+    total_alloc_size_bytes =
+            (db_size * sizeof(*ucs_stats_context.aggrgt_cnt_names_db));
+    ucs_stats_context.aggrgt_cnt_names_db = ucs_realloc(
+            ucs_stats_context.aggrgt_cnt_names_db, total_alloc_size_bytes,
+            (const char*)"stats_cnt_name_db");
```

Comment:
use ucs_array for aggrgt_cnt_names_db

Reply:
aggrgt_cnt_names_db is returned the the user for processing. I think the user of the API would much prefer to get a regular array which can be dereferenced with operator [].

Reply:
you can maintain it as ucs_array but still return a pointer to the first element in this array in the API function

Reply:
Fixed.